### PR TITLE
[2.13.x] DDF-04316 Add better logging for failed automatic logout

### PR DESCRIPTION
**Pull request by tyler30clemens**
#### What does this PR do?
This PR adds better logging to inform the user that they are not actually logged out after a failed automatic logout attempt. This occurs about 5% of the time, and the other times, it returns a successful logout. 

#### Who is reviewing it? 
@austinsteffes 
@ahoffer 

#### Select relevant component teams: 
@codice/security 

#### Ask 2 committers to review/merge the PR and tag them here.
@brjeter 
@stustison 

#### How should this be tested?
1. Install and login through the UI
2. Wait and be idle on the UI until session expiring countdown executes
3. Have chance of being redirected to a successful logout, or `Unable to logout. Please try again.`

#### Any background context you want to provide?
It was discovered that there was a null pointer exception from a race condition. Due to being extremely rare to reproduce, it was decided to just better log that the user is not logged out. 

#### What are the relevant tickets?
For GH Issues:
Work for: #4316 

#### Checklist:
- [ ] Documentation Updated
- [ ] Update / Add Threat Dragon models
- [ ] Update / Add Unit Tests
- [ ] Update / Add Integration Tests

#### Notes on Review Process
Please see [Notes on Review Process](https://codice.atlassian.net/wiki/spaces/DDF/pages/71946981/Pull+Request+Guidelines) for further guidance on requirements for merging and abbreviated reviews. 

#### Review Comment Legend:
- ✏️ (Pencil) This comment is a nitpick or style suggestion, no action required for approval. This comment should provide a suggestion either as an in line code snippet or a gist. 
- ❓ (Question Mark) This comment is to gain a clearer understanding of design or code choices, clarification is required but action may not be necessary for approval.
- ❗ (Exclamation Mark) This comment is critical and requires clarification or action before approval.

### CODE REVIEW

File: platform/security/idp/security-idp-client/src/main/java/org/codice/ddf/security/idp/client/LogoutRequestService.java [hunk 1]
```diff
@@ -180,10 +179,15 @@ public Response sendLogoutRequest(@QueryParam("EncryptedNameIdTime") String encr
           return buildLogoutResponse(msg);
         }
 
+        SecurityToken IdpSecToken = getIdpSecurityToken();
```
Comment (brjeter) [inline platform/security/idp/security-idp-client/src/main/java/org/codice/ddf/security/idp/client/LogoutRequestService.java]: ✏️ Best practice is for the first letter to be lower-case, so it should be `idpSecToken`. (this also applies to the other 2 similar usages below in this file)

File: platform/security/idp/security-idp-client/src/main/java/org/codice/ddf/security/idp/client/LogoutRequestService.java [hunk 2]
```diff
@@ -523,17 +526,22 @@ private void logout() {
   }
 
   private void logSecurityAuditRole() {
+    SecurityToken IdpSecToken = getIdpSecurityToken();
     if (shouldAuditSubject()) {
       SecurityLogger.audit(
           "Subject with admin privileges has logged out: {}",
-          getIdpSecurityAssertion().getPrincipal().getName());
+          new SecurityAssertionImpl(IdpSecToken).getPrincipal().getName());
```
Comment (brjeter) [inline platform/security/idp/security-idp-client/src/main/java/org/codice/ddf/security/idp/client/LogoutRequestService.java]: ❗️ We should check that `IdpSecToken` isn't `null` here as well.

File: platform/security/idp/security-idp-client/src/main/java/org/codice/ddf/security/idp/client/LogoutRequestService.java [hunk 3]
```diff
@@ -523,17 +526,22 @@ private void logout() {
...
     }
   }
 
   private boolean shouldAuditSubject() {
+    SecurityToken IdpSecToken = getIdpSecurityToken();
```
Comment (brjeter) [inline platform/security/idp/security-idp-client/src/main/java/org/codice/ddf/security/idp/client/LogoutRequestService.java]: ✏️ This is technically unrelated to your PR, but thinking we should have the `logSecurityAuditRole` method just pass the token into this method since that's the only usage of this method. Otherwise, we're getting the token twice in a row.

File: ui/packages/security-servlet-logout/src/main/resources/logout-response.html [hunk 4]
```diff
@@ -27,9 +27,8 @@
 <body class="full-width logout-response-body">
 <div class="full-width logout-response-container">
     <div class="logout-msg">
-        <h2>You are now signed out.</h2>
+        <h2 id="extramessage"></h2>
```
Comment (brjeter) [inline ui/packages/security-servlet-logout/src/main/resources/logout-response.html]: I'm surprised that this whole time we've been saying "You are now signed out." in all the various failure scenarios. Good catch 👍

**Comment by austinsteffes** [comment]
hero successful 👍 👌

File: platform/security/idp/security-idp-client/src/main/java/org/codice/ddf/security/idp/client/LogoutRequestService.java [hunk 5]
```diff
@@ -523,17 +526,23 @@ private void logout() {
   }
 
   private void logSecurityAuditRole() {
-    if (shouldAuditSubject()) {
-      SecurityLogger.audit(
-          "Subject with admin privileges has logged out: {}",
-          getIdpSecurityAssertion().getPrincipal().getName());
+    SecurityToken idpSecToken = getIdpSecurityToken();
+    if (idpSecToken != null) {
```
Comment (ahoffer) [inline platform/security/idp/security-idp-client/src/main/java/org/codice/ddf/security/idp/client/LogoutRequestService.java]: Twice we check to see if idpSpecToken is null. I'd like to see that encapsulated in a method with a name meaningful name. I don't know enough about the issue to suggest a good name -- maybe "isSessionMissingToken" or "requestHasToken"?

File: platform/security/idp/security-idp-client/src/main/java/org/codice/ddf/security/idp/client/LogoutRequestService.java [hunk 6]
```diff
@@ -180,10 +179,15 @@ public Response sendLogoutRequest(@QueryParam("EncryptedNameIdTime") String encr
           return buildLogoutResponse(msg);
         }
 
+        SecurityToken idpSecToken = getIdpSecurityToken();
```
Comment (ahoffer) [inline platform/security/idp/security-idp-client/src/main/java/org/codice/ddf/security/idp/client/LogoutRequestService.java]: Why are local variables used to hold the token? Why not call `getIdpSecurityToken` when the token is needed?

File: platform/security/idp/security-idp-client/src/main/java/org/codice/ddf/security/idp/client/LogoutRequestService.java [hunk 7]
```diff
@@ -523,17 +526,23 @@ private void logout() {
...
+      if (shouldAuditSubject(idpSecToken)) {
+        SecurityLogger.audit(
+            "Subject with admin privileges has logged out: {}",
+            new SecurityAssertionImpl(idpSecToken).getPrincipal().getName());
+      }
     }
   }
 
-  private boolean shouldAuditSubject() {
+  private boolean shouldAuditSubject(SecurityToken idpSecToken) {
```
Comment (ahoffer) [inline platform/security/idp/security-idp-client/src/main/java/org/codice/ddf/security/idp/client/LogoutRequestService.java]: The original version of the method did not take an argument. Why the change?

Comment (tyler30clemens) [reply platform/security/idp/security-idp-client/src/main/java/org/codice/ddf/security/idp/client/LogoutRequestService.java]: With the slight refactor, this leads to not having to call `getIdpSecurtyToken` again, when it already knows it from the only place it comes from and is being called (`logSecurityAuiditRole`)

File: platform/security/idp/security-idp-client/src/main/java/org/codice/ddf/security/idp/client/LogoutRequestService.java (hunk 6 above)
Comment (tyler30clemens) [reply platform/security/idp/security-idp-client/src/main/java/org/codice/ddf/security/idp/client/LogoutRequestService.java]: I used a local variable because we want to check if it is null first, but then we use that variable again later. This leads to avoid calling `getIdpSecurityToken` again.

File: platform/security/idp/security-idp-client/src/main/java/org/codice/ddf/security/idp/client/LogoutRequestService.java (hunk 7 above)
Comment (ahoffer) [reply platform/security/idp/security-idp-client/src/main/java/org/codice/ddf/security/idp/client/LogoutRequestService.java]: It gives the impression that different token can be passed in. If the same token (i.e. whatever `getIdpSecurityToken()` returns) is always what method wants, then it should be provided by the class as part of the class's internal state. If `getIdSecurityToken` is expensive in terms of resources it can be cached as a field on the object.

File: platform/security/idp/security-idp-client/src/main/java/org/codice/ddf/security/idp/client/LogoutRequestService.java (hunk 5 above)
Comment (tyler30clemens) [reply platform/security/idp/security-idp-client/src/main/java/org/codice/ddf/security/idp/client/LogoutRequestService.java]: I personally like how it is now because it seems a bit more straightforward, it is only twice, and they are handled differently in the sense that one of them simply checks to see if we should continue and the other returns a `buildLogoutResponse`

File: platform/security/idp/security-idp-client/src/main/java/org/codice/ddf/security/idp/client/LogoutRequestService.java (hunk 7 above)
Comment (tyler30clemens) [reply platform/security/idp/security-idp-client/src/main/java/org/codice/ddf/security/idp/client/LogoutRequestService.java]: This was a suggestion in a previous comment that made sense to me. Since it is only being called there and it is trying to use the same token, it seems reasonable to pass it in as a parameter.

Comment (tyler30clemens) [reply platform/security/idp/security-idp-client/src/main/java/org/codice/ddf/security/idp/client/LogoutRequestService.java]: After further investigation, `logSecurityAuditSubject` and `shouldAuditSubject` were refactored out as helper methods for what looks like just to make the `logout()` method smaller and easier to follow. It previously used the same `securityToken` throughout the process. Unless there was a conscious effort to purposefully call and get that `securtyToken` multiple separate times, I see no issue in using it as a parameter.

Comment (brjeter) [reply platform/security/idp/security-idp-client/src/main/java/org/codice/ddf/security/idp/client/LogoutRequestService.java]: I had seen that the token is retrieved multiple times as well. Not sure if that was intentional, but I would guess that it was. I think this is fine as is. I don't think we want to be caching any security token because the design of this service is that it gets the user information as an input and logs them out. If we start saving user information as state, we run into concurrency problems.

File: platform/security/idp/security-idp-client/src/main/java/org/codice/ddf/security/idp/client/LogoutRequestService.java (hunk 5 above)
Comment (ahoffer) [reply platform/security/idp/security-idp-client/src/main/java/org/codice/ddf/security/idp/client/LogoutRequestService.java]: What does it mean when the token is null?

File: platform/security/idp/security-idp-client/src/main/java/org/codice/ddf/security/idp/client/LogoutRequestService.java (hunk 6 above)
Comment (ahoffer) [reply platform/security/idp/security-idp-client/src/main/java/org/codice/ddf/security/idp/client/LogoutRequestService.java]: What are the negative consequences of calling `getSecurityToken` again?

Comment (tyler30clemens) [reply platform/security/idp/security-idp-client/src/main/java/org/codice/ddf/security/idp/client/LogoutRequestService.java]: Calling `getSecurityToken` calls `getTokenHolder` which calls `getOrCreateSession` which has the potential to try and create a new session. This could be a problem in this case when trying to handle the correct session and assertions.

File: platform/security/idp/security-idp-client/src/main/java/org/codice/ddf/security/idp/client/LogoutRequestService.java (hunk 5 above)
Comment (tyler30clemens) [reply platform/security/idp/security-idp-client/src/main/java/org/codice/ddf/security/idp/client/LogoutRequestService.java]: If the token is null it could mean that the saml assertion was removed from the session. The previous code was calling this in a way that lead to a null pointer exception.

File: platform/security/idp/security-idp-client/src/main/java/org/codice/ddf/security/idp/client/LogoutRequestService.java (hunk 6 above)
Comment (bakejeyner) [reply platform/security/idp/security-idp-client/src/main/java/org/codice/ddf/security/idp/client/LogoutRequestService.java]: I personally prefer having a local variable for the token here, since we use the token twice, and don't need the unnecessary (and probably little) overhead of retrieving the token twice.

File: platform/security/idp/security-idp-client/src/main/java/org/codice/ddf/security/idp/client/LogoutRequestService.java (hunk 5 above)
Comment (bakejeyner) [reply platform/security/idp/security-idp-client/src/main/java/org/codice/ddf/security/idp/client/LogoutRequestService.java]: I think both ideas are great!

However, what would persuade me that encapsulating would be the way to go here, is if the condition we are checking was more complex. Since we are only checking that `idpSecToken` is null, that is simple enough to have inline.

File: platform/security/idp/security-idp-client/src/main/java/org/codice/ddf/security/idp/client/LogoutRequestService.java (hunk 7 above)
Comment (bakejeyner) [reply platform/security/idp/security-idp-client/src/main/java/org/codice/ddf/security/idp/client/LogoutRequestService.java]: I agree with @brjeter's comment. The second we start storing and operating on state in this class, concurrency issues could pop up. I would argue that keeping the code as is is more robust.

**Comment by brjeter** [comment]
Talked offline and decided to merge despite the above comments
